cmds/gzip/gzip.go

```diff
@@ -0,0 +1,436 @@
+	_ = pflag.CommandLine.MarkHidden("seven")
+	_ = pflag.CommandLine.MarkHidden("eight")
+
+	pflag.Parse()
+
+	var err error
+	o.level, err = parseLevels(levels)
+	return err
+
+}
+
+// parseLevels loops through a []bool and returns the index of the element
+// thats true. If more than one element is true return an error. If no
+// element is true, return the constant pgzip.DefaultCompression.
+func parseLevels(levels []bool) (int, error) {
+
+	var level int
+
+	for i, l := range levels {
+		if l && level != 0 {
+			return 0, fmt.Errorf("Multiple compression levels specified")
+		} else if l {
+			level = i
+		}
+	}
+
+	if level == 0 {
+		level = pgzip.DefaultCompression
+	}
+	return level, nil
+}
+
+// file is a file path to be compressed or decompressed.
+type file struct {
+	path    string
+	options *options
+}
+
+// outputPath removes the path suffix on decompress and adds it on compress.
+func (f *file) outputPath() string {
+	if f.options.decompress {
+		return f.path[:len(f.path)-len(f.options.suffix)]
+	}
+	return f.path + f.options.suffix
+}
+
+// checkPath validates the input file path. Checking on compression
+// if the path has the correct suffix, and on decompression checks
+// that it doesn't have the suffix. Allows override by force option.
+func (f *file) checkPath() error {
+	if !f.options.force {
+		if f.options.decompress {
+			if !strings.HasSuffix(f.path, f.options.suffix) {
+				return fmt.Errorf("%s does not have %s suffix", f.path, f.options.suffix)
+			}
+		} else {
+			if strings.HasSuffix(f.path, f.options.suffix) {
+				return fmt.Errorf("%s already has %s suffix", f.path, f.options.suffix)
+			}
+		}
+	}
+	_, err := os.Stat(f.path)
+	if os.IsNotExist(err) {
+		return fmt.Errorf("%s does not exist", f.path)
+	} else if os.IsPermission(err) {
+		return fmt.Errorf("%s permission denied", f.path)
+	}
+	return f.checkOutPath()
+}
+
+// checkOutPath checks if output is attempting to write binary to the stdout.
+// Check if output path already exists. Allow override via force option.
+func (f *file) checkOutPath() error {
+	if f.options.stdout {
+		if !f.options.decompress && !f.options.force {
+			return errors.New("trying to write compressed data to a terminal (use -f to force)")
+		}
+		return nil
+	}
+	_, err := os.Stat(f.outputPath())
+	if !os.IsNotExist(err) && !f.options.force {
+		return fmt.Errorf("%s already exist", f.outputPath())
+	} else if os.IsPermission(err) {
+		return fmt.Errorf("%s permission denied", f.outputPath())
+	}
+	return nil
+}
+
+// Removing input path. Override with keep option. Ignored if output is stdout.
+func (f *file) cleanup() error {
+	if !f.options.keep && !f.options.stdout {
+		return os.Remove(f.path)
+	}
+	return nil
+}
+
+func (f *file) process() error {
+
+	i, err := os.Open(f.path)
+	if err != nil {
+		return err
+	}
+
+	var o *os.File
+	if f.options.stdout {
+		o = os.Stdout
+	} else {
+		if o, err = os.Create(f.outputPath()); err != nil {
+			log.Fatal(err)
+		}
+	}
```

Comment:
Does GNU gzip check Close return errors? Few things will actually generate an error on Close, so it'd be fine to just defer the Close here and above.

Reply:
Especially because the way it currently is, you don't close on errors.